### PR TITLE
PROJ: add new versions, improve tiff patches

### PR DESCRIPTION
**Pull request by danlipsa**
This fixes a proj compilation error from paraview

### CODE REVIEW

**Comment by danlipsa** [comment]
@johnwparent Please review

**Comment by johnwparent** [comment]
@spackbot fix style

**Comment by danlipsa** [comment]
Great. Thanks. This is not ready yet, as it fails on compile on finding some tiff header.

File: var/spack/repos/builtin/packages/proj/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/proj/package.py]: Has this patch been submitted upstream? Would like to permanently fix this in future releases.

File: var/spack/repos/builtin/packages/proj/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/proj/package.py:162]: Should we add another variant and make `+projsync` require `+curl`?

File: var/spack/repos/builtin/packages/proj/package.py
Comment (johnwparent) [reply var/spack/repos/builtin/packages/proj/package.py]: @danlipsa if/when you do upstream this, ping me, I also have a few TIFF related patches we should coordinate on.

Comment (danlipsa) [reply var/spack/repos/builtin/packages/proj/package.py]: They already have something similar in 9.2 so no up-streaming is needed in this case.

**Comment by danlipsa** [comment]
We need a cmake fix for the failing CI test:
https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9300

**Comment by danlipsa** [comment]
@johnwparent ci/gitlab-ci seems to be stuck. Is there a way to restart it? Thanks!

**Comment by johnwparent** [comment]
> @johnwparent ci/gitlab-ci seems to be stuck. Is there a way to restart it? Thanks!

@danlipsa AFAICT, it still seems to actually be running? A job was complete and another was queued just a few moments ago (according to Gitlab). There are some seemingly legitimate failures building Paraview as well.

File: var/spack/repos/builtin/packages/proj/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/proj/package.py]:
```suggestion
    variant("projsync", default=False, description="Build projsync utility", when="+curl")
```

**Comment by adamjstewart** [comment]
Can you resolve the merge conflicts?

**Review by adamjstewart — DISMISSED** [review]
Would be nice to add comments pointing to the upstream version of each patch so future me knows that these are legit and have been accepted upstream

**Review by adamjstewart — DISMISSED** [review]

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Comment by danlipsa** [comment]
@johnwparent @adamjstewart Is this ready? I don't have merge permissions. Errors seem unrelated unless I am missing something.

**Comment by adamjstewart** [comment]
Can't be merged until tests pass. I agree it looks unrelated though. Do any of the failing packages have maintainers who can take a look? I'm away from my computer.

File: var/spack/repos/builtin/packages/cmake/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/cmake/package.py]: Doesn't this have the same issue as https://github.com/spack/spack/pull/42500 ?

**Review by bernhardkaindl — DISMISSED** [review]
Built with build_system=cmake manually:
- It used cmake@3.27.9 instead of cmake@3.28.3.
- `bin/spack install -v proj ^cmake@3.28.3` was successful.

BTW: `bin/spack install proj build_system=autotools` fell back to proj@8.2.1, and failed (with gcc-11): relocation R_X86_64_PC32 against symbol `_ZN5osgeo4proj8metadata10Identifier4EPSGB5cxx11E' can not be used when making a shared object; recompile with -fPIC

Rebased on top of the latest develop branch using GitHub.
Lets find out if the CI issue still persists with it, I guess.

**Comment by adamjstewart** [comment]
@spackbot run pipeline

File: var/spack/repos/builtin/packages/cmake/package.py
Comment (johnwparent) [reply var/spack/repos/builtin/packages/cmake/package.py]: Yes it will/currently does, thank you @alalazo.

**Comment by johnwparent** [comment]
@bernhardkaindl @adamjstewart @danlipsa 
As @alalazo pointed out, this PR will not pass Gitlab CI with these CMake package changes. They are not unrelated, 3.28's FindCUDA overhaul breaks `kokkos-kernel`, for more info see some of the other outstanding CMake upgrade PRs.
A stopgap or real solution for that issue needs to be merged before the changes in this PR can land.

**Comment by bernhardkaindl** [comment]
Hello John, thanks for the more clear update.

It looks like you know best what the issue with CMake, FindCUDA and kokkos-kernelnis. I've not seen any clear description or analysis what the problem with CMake 3.28.3 is, so could you spell out the problem and describe it or point to the problem?

Would it be possible that a future version of CMake can fix the Problem? Can you explain the issue or is there a description of it, so a way "out of it" can be found?

**Comment by johnwparent** [comment]
> It looks like you know best what the issue with CMake, FindCUDA and kokkos-kernelnis. I've not seen any clear description or analysis what the problem with CMake 3.28.3 is, so could you spell out the problem and describe it or point to the problem?

Post 3.28, `kokkos-kernel` fails make due to some invalid target expressions. I am not 100% exactly what's causing this, but I'm fairly confident it's related to FindCUDA after my initial investigation and talking to some of the other CMake folks.

From the error log:
```
CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0111 is not set: An imported target missing its location property
  fails during generation.  Run "cmake --help-policy CMP0111" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.
  IMPORTED_LOCATION not set for imported target "CUDA::cuda_driver"
  configuration "Release".
This warning is for project developers.  Use -Wno-dev to suppress it.
```
This relates to the refactored `FindCUDA.cmake` and is likely the cause of the improper target expression in make.


> Would it be possible that a future version of CMake can fix the Problem? Can you explain the issue or is there a description of it, so a way "out of it" can be found?

The quick solution is to merge my PR adding the conflict to `kokkos-kernel` for CMake's newer than 3.28 (inclusive) and also adding CMake 3.28.4, and then merge this. 
The real solution is to determine which CMake (FindCUDA or `kokkos-kernel`) is causing this and patch it. That will take a bit longer as I'm not sure which yet, but it's under active investigation, see the 3.28.4 version bump PR.

**Review by adamjstewart — APPROVED** [review]

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Now that #43780 is merged, I think we want to remove the patch and only keep the variant controls.

**Comment by danlipsa** [comment]
> Now that #43780 is merged, I think we want to remove the patch and only keep the variant controls.

Seems that my patch does something slightly different. Let me try to compile paraview with my patched commented out to see what happens.

**Comment by scottwittenburg** [comment]
@danlipsa Can you rebase this on #44279 after that gets merged?  That will fix the issue with the overzealous pruning of the DAG, and allow your `proj` changes to be built in gitlab.

**Comment by danlipsa** [comment]
Depends on https://github.com/spack/spack/pull/43723

File: var/spack/repos/builtin/packages/proj/package.py
Comment (johnwparent) [inline var/spack/repos/builtin/packages/proj/package.py]: From https://github.com/spack/spack/pull/44367 Proj 7 does not work with this patch, we should update the version constraints here to reflect that (and not revert that merged change)

**Review by johnwparent — CHANGES_REQUESTED** [review]

**Comment by Chrismarsh** [comment]
I took a look, and there are differences in the `proj@7` cmake that the `@8` patch looks for. I have created a new patch for `@7` as all of `proj@7` appears to use the same TIFF target.  


https://github.com/spack/spack/pull/44535

File: var/spack/repos/builtin/packages/proj/package.py
Comment (Chrismarsh) [reply var/spack/repos/builtin/packages/proj/package.py]: I have created a new patch in https://github.com/spack/spack/pull/44535 to target @7

File: var/spack/repos/builtin/packages/proj/package.py
Comment (johnwparent) [inline var/spack/repos/builtin/packages/proj/package.py]: The patch on line 95 of this file already covers this exact range of changes and because of the open ended constraint, is also applied when this is applied, so this patch no longer works. We need to make the constraints above tighter.

**Review by johnwparent — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/proj/package.py
Comment (danlipsa) [reply var/spack/repos/builtin/packages/proj/package.py]: Indeed. Good catch! I pushed a patch.

**Review by johnwparent — DISMISSED** [review]
Looks good, thanks Dan

**Comment by johnwparent** [comment]
@alalazo this requires you to approve before we can merge

**Review by adamjstewart — COMMENTED** [review]
Will try to review this over the weekend

**Comment by adamjstewart** [comment]
Didn't have much luck with this:
```console
> spack patch proj@9.1.0 build_system=cmake ^cmake@3.28
==> Using cached archive: /Users/Adam/spack/var/spack/cache/_source-cache/archive/81/81b2239b94cad0886222cde4f53cb49d34905aad2a1317244a0c30a553db2315.tar.gz
==> Using cached archive: /Users/Adam/spack/var/spack/cache/_source-cache/archive/f1/f1e5e42ba15426d01d1970be727af77ac9b88c472215497a5a433d0a16dd105b.tar.gz
==> Using cached archive: /Users/Adam/spack/var/spack/cache/_source-cache/archive/dc/dc620ff1bbcc0ef4130d53a40a8693a1e2e72ebf83bd6289f1139d0f1aad2a40
==> Moving resource stage
	source: /var/folders/jv/cgkfvslj6nq1l7cw0c8c_8gm0000gn/T/Adam/spack-stage/resource-proj-data-54valjeojifouh4hayusw4bdaj2qhu3s/spack-src/
	destination: /var/folders/jv/cgkfvslj6nq1l7cw0c8c_8gm0000gn/T/Adam/spack-stage/spack-stage-proj-9.1.0-54valjeojifouh4hayusw4bdaj2qhu3s/spack-src/share/proj
==> Applied patch https://github.com/OSGeo/PROJ/commit/3f38a67a354a3a1e5cca97793b9a43860c380d95.patch?full_index=1
==> Applied patch /Users/Adam/spack/var/spack/repos/builtin/packages/proj/proj-8.1-cmake-3.29-new-tiff-interface.patch
1 out of 1 hunks failed--saving rejects to 'src/lib_proj.cmake.rej'
==> Patch /Users/Adam/spack/var/spack/repos/builtin/packages/proj/tiff_target.patch failed.
==> Error: Command exited with status 1:
'/usr/bin/patch' '-s' '-p' '1' '-i' '/Users/Adam/spack/var/spack/repos/builtin/packages/proj/tiff_target.patch' '-d' '.'
```
Let me see if I can merge all of these patches.

File: var/spack/repos/builtin/packages/proj/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/proj/package.py:162]: This is the last remaining thing I'm unsure of. The build failed for you when `~curl` because it tried building projsync? Is that something we can fix upstream?

**Comment by adamjstewart** [comment]
Tested that all versions now patch. Also tried to build all versions. PROJ 7 fails to build (possibly due to my patch):
```
CMake Error at src/lib_proj.cmake:428 (target_link_libraries):
  The plain signature for target_link_libraries has already been used with
  the target "proj".  All uses of target_link_libraries with a target must be
  either all-keyword or all-plain.

  The uses of the plain signature are here:

   * src/lib_proj.cmake:406 (target_link_libraries)
   * src/lib_proj.cmake:410 (target_link_libraries)
   * src/lib_proj.cmake:418 (target_link_libraries)

Call Stack (most recent call first):
  src/CMakeLists.txt:3 (include)


CMake Error at src/lib_proj.cmake:435 (target_link_libraries):
  The keyword signature for target_link_libraries has already been used with
  the target "proj".  All uses of target_link_libraries with a target must be
  either all-keyword or all-plain.

  The uses of the keyword signature are here:

   * src/lib_proj.cmake:428 (target_link_libraries)

Call Stack (most recent call first): 
  src/CMakeLists.txt:3 (include)
```
PROJ 4.8 and older also fail to build (missing resource?):
```
make[2]: *** [Makefile:254: install-pkgdataDATA] Error 1
make[2]: *** Waiting for unfinished jobs....
nad2nad NADCON source files not present
make[2]: Leaving directory '/private/var/folders/jv/cgkfvslj6nq1l7cw0c8c_8gm0000gn/T/Adam/spack-stage/spack-stage-proj-4.8.0-zz4iuix3zswbpyi5pfko53r26og3vxft/spack-src/nad'
make[1]: *** [Makefile:323: install-am] Error 2
make[1]: Leaving directory '/private/var/folders/jv/cgkfvslj6nq1l7cw0c8c_8gm0000gn/T/Adam/spack-stage/spack-stage-proj-4.8.0-zz4iuix3zswbpyi5pfko53r26og3vxft/spack-src/nad'
make: *** [Makefile:327: install-recursive] Error 1
```
I think we should fix my patch for PROJ 7 and deprecate PROJ 4.8 and older.

File: var/spack/repos/builtin/packages/proj/package.py
Comment (danlipsa) [reply var/spack/repos/builtin/packages/proj/package.py:162]: Yes, having a projsync variant that requires curl would match how the cmakelists.txt is structured.

Comment (danlipsa) [reply var/spack/repos/builtin/packages/proj/package.py:162]: > The build failed for you when ~curl because it tried building projsync? Is that something we can fix upstream?

I don't think so. I think projsync is implemented by using curl.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/proj/package.py:162]: Yes, but there could (should) be something in CMakeLists.txt that automatically disables projsync if curl support is disabled. My concern is that there are hundreds of variants we could add, but it doesn't seem necessary if it's only designed to avoid a build error. Trying to keep the package simple.

**Comment by adamjstewart** [comment]
Everything except PROJ 7.0.0 builds now for me (7.0.1 and 7.2.1 included). 7.0.0 fails to build even without the patch though. I think it's an issue with newer compiler versions. So we just need to decide what to do about projsync and this PR should be good to go.

File: var/spack/repos/builtin/packages/proj/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/proj/package.py:162]: Okay, I can reproduce the error when I disable curl.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/proj/package.py:162]: Maybe it's fine as is. We could also add an `+apps` variant to control all of the apps (not just projsync). But let's wait until someone actually needs this functionality.

**Review by adamjstewart — APPROVED** [review]
Will wait a few days for all interested parties to test this, but I'm pretty confident in this solution now. Hopefully this is the final proj+tiff patch PR for a while.

**Comment by danlipsa** [comment]
@adamjstewart Is this ready to be merged?

**Pull request merged**
